RfC Make shipping rates taxable like line items or shipments

mamhoff opened the pull request:
The taxation code for shipping rates is distributed over `Spree::ShippingRate`, `Spree::Stock::Estimator` and the worst parts of `Spree::TaxRate`. The aim of the taxation refactoring I'm doing is to isolate taxation code to as few moving parts as possible. As discussed on Slack, I'm introducing a spec for taxable items so that I can easily test whether a shipment can possibly be tax adjusted using the tax adjustment mechanism we already have: `Spree::Taxrate.adjust`.

cbrunsdon commented:
On first pass this looks like a fantastic step forward. Nice call. Will review in depth when I get a minute.

mamhoff commented:
I think this works for a new shop. For existing shipping rates, this is a breaking change. However, things are mitigated by the fact that all order-relevant data is actually stored on the shipments.

Do we need to migrate? If so: How would we do that?

jhawthorn commented:
I really like this and I think it's a good direction, but I have some concerns we should think about when reviewing.

As Martin points out, we'll need some sort of migration strategy to move over older shipping rates.

This is the first time we've calculated and stored tax rates on an item which was not included in the order's total proper. This could have issues if a site did any reporting by counting adjustments through their `order_id`. Maybe we should consider having `order_id=nil` for these objects.

To be clear, I think this is a good step forward, I just want extra caution as always with anything tax-ralted.

mamhoff commented:
As for the old shipping rates: We have to consider that old shipping rates might have been taxed using different tax rates, as they might have been deleted or updated in the meantime. This opens up another issue I've been hoping to be able to defer: How can we transparently handle changing tax rates? I.e.: A country raises their Sales tax by the beginning of, say, 2016... Ideally I'd have tax rates be immutable, and some kind of `applicable_from` and `applicable_until` field to filter for those that are valid right now.

For deleted tax rates, which are `paranoid`, we can backfill the tax rates' applicability dates using `created_at` and `deleted_at`. For updated ones, we're out of luck (but in these cases the current code also applies the updated ones now). 

It'd be a lot simpler to move that migration to the end of this branch, and not try stuffing it into the current situation, as there's lots of places where that might be necessary.

mamhoff commented:
As for the `order_id = nil` question: I'm hesitant on that one, as it would mean that shipping rate taxes would not be recalculated in the order updater. 

Given that the `all_adjustments` association only made it into Solidus in October last year, I would doubt people use it to generate reports - and if they did, I'd wonder what sense it would make to group tax adjustments and (potentially non-eligible) promotion adjustments in one report.

mamhoff commented:
Rebased on current master, including the new shipping specs.

mamhoff commented:
I do not think we need to migrate, even for existing shops: 

Shipping rate adjustments actually do not show up anywhere except for the select box where customers choose a shipping method in the shipment step. I'm not sure we need a migration for this: If the order is `complete`, we don't show shipping options anymore; if it's not `complete` and the customer chooses the shipping method again, he'll have gone through a state transition on the order where shipments and shipping rates get re-calculated. 

Setting `order_id` to `nil`: 

This would require patching `Spree::Adjustment` to not validate the presence of the order id under some circumstances. I'm now using the `eligible` bit, which indicates whether an adjustment actually counts toward the adjustment total of an order, and I ask the item being tax adjusted whether it's eligible itself when I set it. See the commit message for 2eb5d0a for more explanations.

mamhoff commented:
In private conversation with @jordan-brough he voiced concerns about further cementing `eligible` adjustments, or using the adjustment system for something other than actually adjusting orders.

Likewise, other members of the community would like to see something like taxation system that is pluggable, and not even necessarily in core. 

https://github.com/solidusio/solidus/pull/862 addresses some of the issues we currently have with shipping rates, and the refactoring of the `display_price` method can be ported to that as well. Furthermore, it's a better step towards isolating Spree's inner workings from the taxation stuff. 

These were supposed to be complementary, but can also be thought of different ways of achieving similar goals. @jordan-brough I'd still like you to look over this though; many thanks for yesterday's conversation!

jordan-brough commented:
@mamhoff thanks for all the details!  I've been looking at this today and am still wrapping my brain around everything.  Will look some more and leave more thoughts tomorrow.

jhawthorn commented:
I'm starting be less terrified by this. Making the shipping rate adjustments eligible=false helps my confidence in safe upgrades a lot.

A couple concerns still:
* `eligible?` on the taxable items isn't the most descriptive. Looking at `line_item.eligible?` the purpose of that method is unclear. Maybe something like `tax_adjustments_eligible?`.
* The adjustments on ShippingRates don't seem to be deleted when they are. (Orphan records added to the DB when refresh_rates is called)
* This adds a lot of queries when calculating shipping rates. Maybe unavoidable.

mamhoff commented:
I'm actually less and less a fan of this. It introduces useless columns (`promo_total` for something that can't be promoted, for example - but really all the totals as they aren't ever used). 

#862 showed me that this is possible entirely without adjustments, at the expense of quite a few database calls. I think I can find a solid middle ground with a new AR model `Spree::Tax::ShippingRateTaxEstimation` which would just have an amount and a tax rate, and has no direct connection to the order (because it doesn't need it). That can also be easily used for migrating old shipping rates. 

Having `ItemAdjustments` do a bunch of re-calculations every time the order is updated IMO doesn't help a lot either. Stay tuned.

mamhoff commented:
Closing this in favor of #904. Much more fun.
